src/som/interpreter/actors/Actor.java

```diff
@@ -190,12 +203,73 @@ protected void appendToMailbox(final EventualMessage msg) {
     mailboxExtension.append(msg);
   }
 
-  protected static void handleBreakPoints(final EventualMessage msg, final WebDebugger dbg) {
-    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED && msg.isMessageReceiverBreakpointSet()) {
+  protected static void handleBreakpointsAndStepping(final EventualMessage msg, final WebDebugger dbg, final Actor actor) {
+    if (!VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+      return;
+    }
+
+    synchronized (actor) {
+
+      if (msg.isStepToReceiver() || msg.isMessageReceiverBreakpointSet() || actor.isStepToNextTurn()) {
+        dbg.prepareSteppingUntilNextRootNode();
+        if (actor.isStepToNextTurn()) { // reset flag
+          actor.setStepToNextTurn(false);
+        }
+      }
+
+      if (msg.isStepToNextTurn()) {
```

Comment:
I did not realize I have it duplicated, the semantic I was thinking was not for this case, now I think it must be only in the actor